### DISCUSSION TITLE
OnRecordAfterCreateRequest for OAuth2 logins

### DISCUSSION THREAD

**Discussion by jkdoshi** [pocketbase/pocketbase#2052] [Q&A] [answered]
I'm trying use app.OnRecordAfterCreateRequest hook. It works perfectly if I create a user manually as admin (using admin UI). I'm using it to create a linked account in another system from the backend.

But I have also enabled login-with-google. When a user effects the creation of their own auth-record by using that feature, the above hook never triggers. I would've thought that the above should trigger. Or maybe there's another hook that I should use?

Any help would be appreciated.

**Comment by ganigeorgiev** [comment, marked as answer]
The `OnRecordAfterCreateRequest` fires only on direct `POST /api/collections/:collections/records` request.

If you want to listen to OAuth2 sign-in/sign-up requests, you can bind to the `OnRecordAfterAuthWithOAuth2Request` hook.

If you want to listen to all user `Record` models create event (including those created programmatically via `app.Dao()`), you can bind to the `OnModelAfterCreate` hook.

**Comment by jkdoshi** [comment]
In `OnRecordAfterAuthWithOAuth2Request` how do I tell if this is the first time user creation/sign-up or subsequent auth/sign-in after creation? The parallel account I'm creating needs to be created only the first time/sign-up, not every time/sign-in.

**Reply by ganigeorgiev** [reply]
Hm, there is no easy way to distinguish OAuth2 sign-in vs sign-up at the moment. One hacky workaround would be to attach also to the `OnRecordBeforeAuthWithOAuth2Request` and check the `e.Record` whether is `nil` or not. If it is `nil` you can set a local/context flag via `e.HttpContext.Set("isNew", e.Record == nil)` and check it from the after hoook via `e.HttpContext.Get("isNew")`.
I'll consider in the next release to add a similar flag directly in the hook event.

But for your case it might be better to attach to the `OnModelAfterCreate` hook since it will be triggered for all user create scenarios (_it is a db/dao event so the main drawback is that you don't have access to the HTTP request context_):

```go
app.OnModelAfterCreate("users").Add(func(e *core.ModelEvent) error {
	record := e.Model.(*models.Record)

	// do something with the created users record...

	return nil
})
```

**Reply by jkdoshi** [reply]
Thank you for the thoughtful and detailed response (as always).

> But for your case it might be better to attach to the `OnModelAfterCreate` hook since it will be triggered for all user create scenarios (_it is a db/dao event so the main drawback is that you don't have access to the HTTP request context_):

And that's exactly why I have to use this hook. I need to access HTTP request context.

What about having a `*Request` hook that triggers with HTTP context when an OAuth accounts get linked to an auth-record?

**Reply by ganigeorgiev** [reply]
> What about having a *Request hook that triggers with HTTP context when an OAuth accounts get linked to an auth-record?

I think it will be confusing because you are not making the account linking request directly (and if the OAuth2 provider is already linked it will be ambiguous whether the hook should be called or not).

I'll consider updating the existing `RecordAuthWithOAuth2Event` struct by adding more fields about the OAuth2 request (aka. whether is sign-in or sign-up, first-time linking, etc).

**Comment by jkdoshi** [comment]
Thank you!

For anyone trying something similar, here's what I ended up doing ...

```go
app.OnModelAfterCreate().Add(func(e *core.ModelEvent) error {
	// e.Model is low-level data without the Record wrapper. So type assert it into a Record and also check that it's an auth record
	if record, ok := e.Model.(*models.Record); ok && record.Collection().IsAuth() {
		log.Printf("Creating Metabase account for: %s (%s)\n", record.Email(), record.Id)
		// ... parallel acount creation code using Record above
	}
	return nil
})

```

The above is very easy and it works for both password logins as well as OAuth2, but slightly less than ideal because I can't piggyback some extra info into the HTTP response body or headers.

**Reply by ganigeorgiev** [reply]
_One minor note: keep in mind that the above will be triggered for any model create (not just users)._

_If you want to restrict it to only your auth collection, you can add tag(s) to the hook (eg. `app.OnModelAfterCreate("users")`) or manually check the record collection inside the hook hander (eg. `record.Collection().Name == "users"`)._

**Comment by jkdoshi** [comment]
Oh yeah. I forgot to include that I'm checking for `record.Collection().IsAuth()` before proceeding. I've updated my sample code above.
